### PR TITLE
Issue#32 V57033H/I/J all column types using Physical Keys

### PR DESCRIPTION
**Pull request by ashtkumar2023**
VTL157033J through N tables have been created. Kindly review.

### CODE REVIEW

**Review by robe070 — CHANGES_REQUESTED** [review]
Names of tables need to include the ticket number and specify the column type being used.
e.g. VTL157033H is
**157033** - Select Generic **Varchar** Physical
So for table I it will be
**157033** - Select Generic **NChar** Physical
etc.
An example was given in Table H. Take a little more consideration and do it correctly the first time. It will be quicker for you.

The physical key is not correct in all the new tables. Again Table H is your example. See how the Varchar short name column is the last part of the key? For each table the particular column type you are testing needs to be part of the key.
image

The Indexes seem to be constructed correctly.

**Comment by robe070** [comment]
I have reviewed the code in `V57033G`. I'm further removing functionality. Delete table `VTL157033M` which operates over `SBIN` and all the code that tests Fields `F157033S` and `F157033S_BIN`.
e.g.
```

* ALPHA + SBIN.
Select Fields(#F157033S) From_File(L157033DS) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
#LF_Func_L157033DS += 1
Endselect

If ((#this_rc = ER) *Or (#this_rc = VE))
#LF_Total_Errors += 1
#Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS' )
Else
Select Fields(#F157033S_SBIN) From_File(L157033DS_SBIN) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
#LF_Func_L157033DS_SBIN += 1
Endselect

If ((#this_rc = ER) *Or (#this_rc = VE))
#LF_Total_Errors += 1
#Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS_SBIN' )
Else
#this_rc := OK
Endif
Endif
```

In `mthValidate_Results` you will find code to test SBIN. remove it:
```
* Alpha & SBIN records should be identical.
If (#LF_Func_L157033DA <> #LF_Func_L157033DS)
#LF_Total_Warnings += 1

#Com_Owner.mthAdd_SubMsg( #this_rc #this_operation ('*** WARNING (Alpha/SBIN) records not matched: (' + #LF_Func_L157033DA.AsString + ', ' + #LF_Func_L157033DS.AsString + ')') )
```

**Comment by ashtkumar2023** [comment]
> For each table the particular column type you are testing needs to be part of the key

> Names of tables need to include the ticket number and specify the column type being used. e.g. VTL157033H is **157033** - Select Generic **Varchar** Physical So for table I it will be **157033** - Select Generic **NChar** Physical etc. An example was given in Table H. Take a little more consideration and do it correctly the first time. It will be quicker for you.
> 
> The physical key is not correct in all the new tables. Again Table H is your example. See how the Varchar short name column is the last part of the key? For each table the particular column type you are testing needs to be part of the key. image
> 
> The Indexes seem to be constructed correctly.

Field F1570331 looks like this : 
image

Field F1570338 looks like this : 
I Understand  description of the Physical key needs to be changed to NChar(100)/String(40) in the field attributes. Is my understanding correct ?
image

**Comment by ashtkumar2023** [comment]
> I have reviewed the code in `V57033G`. I'm further removing functionality. Delete table `VTL157033M` which operates over `SBIN` and all the code that tests Fields `F157033S` and `F157033S_BIN`. e.g.
> 
> ```
> 
> * ALPHA + SBIN.
> Select Fields(#F157033S) From_File(L157033DS) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
> #LF_Func_L157033DS += 1
> Endselect
> 
> If ((#this_rc = ER) *Or (#this_rc = VE))
> #LF_Total_Errors += 1
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS' )
> Else
> Select Fields(#F157033S_SBIN) From_File(L157033DS_SBIN) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
> #LF_Func_L157033DS_SBIN += 1
> Endselect
> 
> If ((#this_rc = ER) *Or (#this_rc = VE))
> #LF_Total_Errors += 1
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS_SBIN' )
> Else
> #this_rc := OK
> Endif
> Endif
> ```
> 
> In `mthValidate_Results` you will find code to test SBIN. remove it:
> 
> ```
> * Alpha & SBIN records should be identical.
> If (#LF_Func_L157033DA <> #LF_Func_L157033DS)
> #LF_Total_Warnings += 1
> 
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation ('*** WARNING (Alpha/SBIN) records not matched: (' + #LF_Func_L157033DA.AsString + ', ' + #LF_Func_L157033DS.AsString + ')') )
> ```
Should I read V57033G as V57033H ?
Ok..I will remove VTL157033M from my  tables and push again .

**Comment by ashtkumar2023** [comment]
> > For each table the particular column type you are testing needs to be part of the key
> 
> > Names of tables need to include the ticket number and specify the column type being used. e.g. VTL157033H is **157033** - Select Generic **Varchar** Physical So for table I it will be **157033** - Select Generic **NChar** Physical etc. An example was given in Table H. Take a little more consideration and do it correctly the first time. It will be quicker for you.
> > The physical key is not correct in all the new tables. Again Table H is your example. See how the Varchar short name column is the last part of the key? For each table the particular column type you are testing needs to be part of the key. image
> > The Indexes seem to be constructed correctly.
> 
> Field F1570331 looks like this : image
> 
> Field F1570338 looks like this : I Understand description of the Physical key needs to be changed to **1570338-NChar(100)/String(40)** in the field attributes. Is my understanding correct ? image

**Comment by ashtkumar2023** [comment]
Kindly  review. PFA verifier report.
[Verifier_Test_Report.txt](https://github.com/lansa/db-regression/files/13932089/Verifier_Test_Report.txt)

**Comment by robe070** [comment]
> I have reviewed the code in `V57033G`. I'm further removing functionality. Delete table `VTL157033M` which operates over `SBIN` and all the code that tests Fields `F157033S` and `F157033S_BIN`. e.g.
> 
> ```
> 
> * ALPHA + SBIN.
> Select Fields(#F157033S) From_File(L157033DS) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
> #LF_Func_L157033DS += 1
> Endselect
> 
> If ((#this_rc = ER) *Or (#this_rc = VE))
> #LF_Total_Errors += 1
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS' )
> Else
> Select Fields(#F157033S_SBIN) From_File(L157033DS_SBIN) With_Key(#this_test_value) Nbr_Keys(*COMPUTE) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
> #LF_Func_L157033DS_SBIN += 1
> Endselect
> 
> If ((#this_rc = ER) *Or (#this_rc = VE))
> #LF_Total_Errors += 1
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation 'On Logical View L157033DS_SBIN' )
> Else
> #this_rc := OK
> Endif
> Endif
> ```
> 
> In `mthValidate_Results` you will find code to test SBIN. remove it:
> 
> ```
> * Alpha & SBIN records should be identical.
> If (#LF_Func_L157033DA <> #LF_Func_L157033DS)
> #LF_Total_Warnings += 1
> 
> #Com_Owner.mthAdd_SubMsg( #this_rc #this_operation ('*** WARNING (Alpha/SBIN) records not matched: (' + #LF_Func_L157033DA.AsString + ', ' + #LF_Func_L157033DS.AsString + ')') )
> ```

The only part of this that you have done is to remove table M. Please do the rest as your next commit. No other changes. Its code removal. I just want to see that tidied up before more changes are done.

Also as part of this next commit, remove all the code that deals with the Reusable Part that is declared on this line and used elsewhere in the code. 
```
Define_Com Class(#R157033A) Name(#o_Query)
```

image

This is in the original ticket - the last line _Remove the calls to the Reusable Part in V57033H_

**Comment by robe070** [comment]
After that commit (and push to the PR so I may easily review it in isolation), fix up Table H testing.

e.g. `L157033D1_VARCHAR` every where needs to change to `L157033H1_VARCHAR`. In the counters (like `LF_Func_L157033D1_VARCHAR`), in the IO commands (like `From_File(L157033D1_VARCHAR`) and in messages (like `'On Logical View L157033D1_VARCHAR'`).

A simple global change will do it.

Commit and push that to the PR too.

**Comment by ashtkumar2023** [comment]
1) Thank you . I am working on the above feedback  2) Can you please provide your comment on this line inside the routine? Mthroutine Name(mthSearch_Generic_By_2_Varchar_Keys)                                             #LF_Key1_And_Key2_WithKey_Counter := 0
Select Fields(#F157033K1) From_File(L157033D1) With_Key(#F1570331 #F1570331_VARCHAR) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
#LF_Key1_And_Key2_WithKey_Counter += 1
Endselect

**Comment by robe070** [comment]
> 1. Thank you . I am working on the above feedback  2) Can you please provide your comment on this line inside the routine? Mthroutine Name(mthSearch_Generic_By_2_Varchar_Keys)                                             #LF_Key1_And_Key2_WithKey_Counter := 0
>    Select Fields(#F157033K1) From_File(L157033D1) With_Key(#F1570331 #F1570331_VARCHAR) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
>    #LF_Key1_And_Key2_WithKey_Counter += 1
>    Endselect

This means that the Physical Key on `VTL157033H` needs to be the same as the Key on the index `L157033D1`. Then the IO is changed from:
```
Select Fields(#F157033K1) From_File(L157033D1) With_Key(#F1570331 #F1570331_VARCHAR) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
```
to
```
Select Fields(#F157033K1) From_File(VTL157033H) With_Key(#F1570331 #F1570331_VARCHAR) Generic(*YES) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
``` 

Likewise for all references to `L157033D1` in that function.

**Comment by ashtkumar2023** [comment]
> _SBIN

F1570331_VARCHAR is not a Primary Key column in VTL157033H.
image

**Comment by robe070** [comment]
> > _SBIN
> 
> F1570331_VARCHAR is not a Primary Key column in VTL157033H. image

What are you trying to say?
"_SBIN?"
????????

"F1570331_VARCHAR is not a Primary Key column in VTL157033H"?

I've told you how to address that. And its NOT like your screen shot. Is the primary key you've selected for `VTL157033H` the same as `L157033D1`? No its not. Fix it.

**Comment by ashtkumar2023** [comment]
Fixing it.

**Comment by ashtkumar2023** [comment]
We could proceed to make the "F1570331_VARCHAR" as primary key.                                                                                       As per the ticket description,
**Short Name** : F1570331, will be replaced by a query over the physical i.e. VTL157033H
**Long Name**:   F1570331_VARCHAR - retained as is over LV
image

**Comment by robe070** [comment]
I showed you this in our video meeting. I've texted you about this. Its trivial to do.

"Is the primary key you've selected for VTL157033H the same as L157033D1? No its not. Fix it."

Here is the key for `L157033D1`
image

That is NOT the same as the Primary key you have shown me in the latest image of `VTL150733H`.

What is different Ashwini?

**Comment by robe070** [comment]
`VTL150733H`
image

**Comment by ashtkumar2023** [comment]
PFA.
[Verifier_Test_Report.txt](https://github.com/lansa/db-regression/files/13952032/Verifier_Test_Report.txt)

File: Partition (WBP)/Functions/V57033H.VT157033.fun.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@
     * This is a copy of V57033D with one LV query replaced by a physical query.
     *
     * Tests single and 2 key generic search of ALL 'string' field types. That is:
-    * Varchar, String, Nchar, Nvarchar, Alpha, Char, Alpha SBIN, Char SBIN.
+    * Varchar, String, Nchar, Nvarchar, Alpha, Char, Char SBIN.
```
Comment (robe070) [inline Partition (WBP)/Functions/V57033H.VT157033.fun.yml]: Char SBIN should be Char SUNI

**Review by robe070 — CHANGES_REQUESTED** [review]
#F157033B and #F157033B_BINARY need to be removed from V57033H.

Remove these 3 commented lines
```

* Remove this test case due to it had already failed in V14 SP1 on IBM i, and no intention
* to make any change to the Native Coding area.
* #Com_Owner.mthSearch_Generic_By_2_SBIN_Keys( #LF_Test_String #LF_Test_String #this_rc )
```

Re: _"Likewise for all references to L157033D1 in that function."_
You have not renamed the counters. Do a global change of L157033D1 to L157033H1.
That indicates you didn't do a global change for the rest. Curious as to why not?

**Comment by ashtkumar2023** [comment]
> F157033B_BINARY

Can I  read #F157033B and #F157033B_BINARY as #F1570332 #F1570332_VARBINARY?

File: Partition (WBP)/Functions/V57033H.VT157033.fun.yml [hunk 2]
```diff
@@ -139,6 +150,7 @@
     #F1570331 #F1570331_VARCHAR #F1570338 #F1570338_NCHAR #F1570339 #F1570339_NVARCHAR #F157033A #F157033A_ALPHA #F157033H #F157033H_CHAR #F157033U #F157033U_SUNI := #LF_Test_Data
     
     Insert Fields(*ALL) To_File(VTL157033H) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
+    Insert Fields(*ALL) To_File(VTL157033I) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
```
Comment (robe070) [inline Partition (WBP)/Functions/V57033H.VT157033.fun.yml:202]: Each table needs separate RC testing.

**Review by robe070 — CHANGES_REQUESTED** [review]
Just the one issue with the INSERT

File: Partition (WBP)/Functions/V57033H.VT157033.fun.yml (hunk 2 above)
Comment (ashtkumar2023) [reply Partition (WBP)/Functions/V57033H.VT157033.fun.yml:202]: Ok..I will fix this ,push and then move to 'J'

Comment (robe070) [reply Partition (WBP)/Functions/V57033H.VT157033.fun.yml:202]: agreed

File: Partition (WBP)/Functions/V57033H.VT157033.fun.yml [hunk 3]
```diff
@@ -159,19 +168,67 @@
...
+    #F1570331 #F1570331_VARCHAR #F1570338 #F1570338_NCHAR #F1570339 #F1570339_NVARCHAR #F157033A #F157033A_ALPHA #F157033H #F157033H_CHAR #F157033U #F157033U_SUNI := #LF_Test_Data
+    
+    Insert Fields(*ALL) To_File(VTL157033K) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
+    If (#this_rc *NE OK)
+    #this_rc := ER
+    Leave
+    Endif
+    End_Loop
+    Endif
+    #Com_Owner.mthAdd_SubMsg( #this_rc 'Insert Data' (#F157033K2.AsString + ' / ' + #LF_Test_Data_Count.AsString + ' records inserted into VTL157033K') )
```
Comment (robe070) [inline Partition (WBP)/Functions/V57033H.VT157033.fun.yml]: All wrong. See main review.

**Review by robe070 — CHANGES_REQUESTED** [review]
Just 1 insert loop. Not 1 loop for each table. Like this:

```
If (#this_rc = OK)
Begin_Loop Using(#LF_i) To(#LF_Test_Data_Count)
Get_Entry Number(#LF_i) From_List(#wLTestData)

#F157033K2 += 1
#F1570331 #F1570331_VARCHAR #F1570338 #F1570338_NCHAR #F1570339 #F1570339_NVARCHAR #F157033A #F157033A_ALPHA #F157033H #F157033H_CHAR #F157033U #F157033U_SUNI := #LF_Test_Data

Insert Fields(*ALL) To_File(VTL157033H) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
If (#this_rc *NE OK)
#this_rc := ER
Leave
Endif

Insert Fields(*ALL) To_File(VTL157033I) Io_Status(#this_rc) Io_Error(*NEXT) Val_Error(*NEXT)
If (#this_rc *NE OK)
#this_rc := ER
Leave
Endif
End_Loop
Endif
```

**Comment by ashtkumar2023** [comment]
> Get_Entry Number(#LF_i) From_List(#wLTestData)

Fixed the above comment.
[Verifier_Test_Report.txt](https://github.com/lansa/db-regression/files/13971440/Verifier_Test_Report.txt)

**Comment by robe070** [comment]
That is correct.

**Comment by robe070** [comment]
Are you finished?

**Comment by ashtkumar2023** [comment]
I committed L and N in  [913ca2e](https://github.com/lansa/db-regression/pull/65/commits/913ca2e06cb503bbbe85e9574df982169e6302c7)

**Comment by robe070** [comment]
Correct. Are you finished?

**Comment by ashtkumar2023** [comment]
I have started working on table 'O'

**Pull request merged**